### ISSUE TITLE
How do NodePort Services work?

### ISSUE THREAD

**Issue by MartiniMoe** [kube-hetzner/kube-hetzner#36]
Hi there,

thank you very much for your awesome help in my last ticket regarding the TLS configuration. It works now :)

Now im wondering, how exposing a service with a NodePort works. If I understand it correctly the firewall and loadbalancer should get configured to forward the port once I create a NodePort Service in my cluster, but that does not happen.

Is my assumption correct? Is NodePort not possible with kube-hetzner?

Thank you very much

**Comment by mysticaltech** [comment]
Hey @MartiniMoe, you just need to open the incoming port on the Hetzner firewall in the project where your Kube cluster lives. As for the load balancer, it does not come into the picture here. NodePort is opening directly on the ports, so no interaction whatsoever with the LB.

Basically, you open the port on the Hetzner Firewall, then create the NodePort, and you should be able to hit the service with the Node IP directly.

It's been a long time I have not done this, but that should be the case. Nowadays, I just use Ingress through the LB for everything.

**Comment by MartiniMoe** [comment]
But then you can't use - for example - the default ports of dedicated game servers, right? Because everything goes over Port 80 or 443, right?

**Comment by mysticaltech** [comment]
Yes, you could, you are not at all limited to HTTP or SSL, you can open any port you wish as a NodePort, open a hole in the firewall, and that's it. 

Now you could also do the same with your Traefik ingress controller, as unlike Nginx it's not limited to the two ports above either, it can do any [TCP or even UDP ports](https://doc.traefik.io/traefik/v2.2/routing/entrypoints/). Again, you never ever touch the LB manually, Traefik has everything it needs thanks to the proxy protocol comms it has with the LB.

**Comment by MartiniMoe** [comment]
Thanks! If I would do it with a Traefik Ingress I would define another entrypoint in templates/traefik_config.yaml.tpl for that port, right?

The thing is I do not understand until now how port 80 and 443 are mapped to the existing  entrypoints web and websecure. Maybe I should just get more into terraform for the moment...

**Comment by mysticaltech** [comment]
Don't worry, it's not about Terraform. Everything happens in [templates/traefik_config.yaml.tpl ](https://github.com/kube-hetzner/kube-hetzner/blob/d722c0e7b642f0b0ab2210e0451f86e0dcce5c4d/templates/traefik_config.yaml.tpl), as you mentioned above. 

Basically, you are dealing with a [HelmChartConfig](https://rancher.com/docs/k3s/latest/en/helm/#customizing-packaged-components-with-helmchartconfig), and you want to add another entrypoint, so find how you do that with the [Traefik Helm Values file,](https://github.com/traefik/traefik-helm-chart/blob/master/traefik/values.yaml) and add those values in the HelmChartConfig template above.

The section you probably want to add/and customize is the `ports` section [here](https://github.com/traefik/traefik-helm-chart/blob/6aae2789613d5cde6767dd1f809ca05f667f3959/traefik/values.yaml#L252). 

In the screenshot, you can see that `ports` is a sister key to `additionalArguments` so do the same, just copy-paste the whole ports section in the HelmChartConfig, and add the entrypoints for your game servers there, on ports other than, the ones already used. Last but not least, add the same additionalArguments for them, to allow proxy protocols.

**Comment by MartiniMoe** [comment]
Awesome, will try that later, thank you so much!

**Comment by mysticaltech** [comment]
Hey @MartiniMoe, will close this, but if you have any more questions, do not hesitate to continue the discussion here.
